### ISSUE TITLE
Investigate nodetemplate ID options

### ISSUE THREAD

**Issue by cloudnautique** [rancher/rancher#26160]
The data migration to support Rancher admins being able to edit them involved a data migration. The result was the IDs changed on the node templates, and clients like the Terraform provider can no longer access the node template by the old IDs. The Terraform provider believes it needs to create a new one, even though one already exists. 

To address this, we need engineering to investigate the following options:
1. Is there something that can be done to make the resource ids reachable post-upgrade?
2. If the ids are manually modified post-upgrade with the pre-upgrade values would that be supportable? (though not ideal)

**Comment by rawmind0** [comment]
Submitted and tested PR https://github.com/terraform-providers/terraform-provider-rancher2/pull/288 to fix issue on terraform provider

**Comment by rmweir** [comment]
Just leaving these notes here per what we have discussed.

* First, we discussed a solution where if a nodetemplate is requested from rancher that has been replaced, then the new template is returned instead if user has RBAC for it. This should not break clients consuming the API.

* The first proposed solution is abnormal and unpredictable design for an API. A more idiomatic solution would be to return a redirect response. However, this would break clients that are not prepared to handle a redirect response.

**Comment by izaac** [comment]
I was able to test the workflow for Node Templates:

Reproduced the issue with the TF provider `master` branch:

`v2.3.2 -> v2.3.3`

Then validated the fix with:

`v2.3.2 -> v2.3.3 -> v2.3.6 -> v2.4.1` using the TF provider `288` branch for testing. 
Working on an scenario to include more Rancher objects at the moment.

**Comment by izaac** [comment]
Tested the following E2E scenario:

1. Deploy rancher server - v2.3.2
2. Using terraform scripts:
3. Create Node Template -T1
4. Deploy a cluster using C1
5. Upgrade rancher server to 2.3.6
6. Using terraform scripts:
7. **Scale up a node pool in cluster C1 that uses T1.**

In the step 7 above `terraform apply` increasing only the node pool `quantity` and nothing else gets the node pool count increased but a new Node Template gets created and this new template is used in the new node in the nodepool.

**Comment by izaac** [comment]
After further investigation and testing with @rmweir we identified I had a binary cached by terraform that was using the latest stable release code (old code without the patch).

We validated my previously [commented scenario](https://github.com/rancher/rancher/issues/26160#issuecomment-607547252) passed for both:

- `v2.3.0` to `v2.3.3`
- `v2.3.2` to `v2.3.6`

**Comment by rmweir** [comment]
We have decided that the best solution for now is:
when a user requests a legacy node template, return the migrated version instead, change the id to match the legacy id.
Example:
User gets `nodetemplates/u-abcd:nt-efgh`
rancher should get note template cattle-global-data:nt-u-abcd-nt-efgh (or w/e new id is) and then change the id on the returned object to `u-abcd:nt-efgh` before returning it to user

**Comment by cbron** [comment]
Merged but waiting for 2.4.3 release to backport to 2.4 branch before moving to test.

**Comment by rmweir** [comment]
Recommended steps to test:
Use an older version of the rancher tf provider (I think 1.8.1 should do), then use it in Rancher 2.3.2 to create a node template. Upgrade to master, then run tf apply. If a new node template is created, the issue is present, if it doesn't, then the issue should be gone.

Should be able to access the node template by its old id, but `kubectl get nodetemplates` should only show the new node template, and not the legacy one. New node template id will be of the form "nt-(user)-(legacy-id)".

**Comment by izaac** [comment]
Terraform provider version v1.8.1
Upgrade from Rancher `v2.3.2` to Rancher `v2.4.2`
Trying to apply the terraform configuration again in Terraform, recreate the node templates as new templates according to `terraform plan` and update the existing node template IDs.

`Plan: 2 to add, 3 to change, 0 to destroy.`

Upgrade from Rancher `v2.3.2` to Rancher `v2.4-head`
Trying to apply the terraform configuration again in Terraform, just update the node templates IDs according to `terraform plan`

`Plan: 0 to add, 3 to change, 0 to destroy.`

After applying the terraform config I still have access through the API to the old nodetemplate ID in the form of:

`user-vrqhv:nt-qngwq`

The new generated ID after the update:

`cattle-global-nt:nt-user-vrqhv-nt-qngwq`

**Comment by izaac** [comment]
### Upgrade from Rancher `v2.3.2` to Rancher `master-head (05/11/2020) 84d937acf` 

Same results as `v2.4-head` describe in my [previous comment](https://github.com/rancher/rancher/issues/26160#issuecomment-626861994).

**Comment by izaac** [comment]
Terraform provider version `v1.8.3` (latest as of 05/11/2020).

### Upgrade from Rancher v2.3.2 to Rancher master-head (05/11/2020) 84d937acf

Same results as master-head described previously.

**Comment by izaac** [comment]
Terraform provider version `v1.8.3` (latest as of 05/11/2020).

### Upgrade from Rancher `v2.3.2` to Rancher `v2.4.2`

This terraform provider version has the fix on the TF provider code. In this case Terraform didn't create new resources as expected and No updates were applied in the node templates ids.

`No changes. Infrastructure is up-to-date.`

**Comment by izaac** [comment]
From Rancher perspective I did an upgrade test and spot checked using node pools that reference the node template IDs.

- Upgrade from `v2.3.2` to `v2.4.3` 

I was able to see the new node template IDs but I wasn't able to access the legacy IDs. API returned 404.

- Upgrade from `v2.3.2` to `v2.4-2948-head` commit id: 65b520517

I was able to see the new node template IDs and I was also able to access the old ID and retrieve the data. The ID returned in the template was the legacy id requested.

Checked increasing and decreasing the scale of node pools.
